TST enable non-CPU device testing via array-api-strict

ogrisel opened the pull request:
This is an early draft PR to attempt to leverage multi device support recently merged in `array-api-strict`: https://github.com/data-apis/array-api-strict/pull/59

We need to wait for a release of `array-api-strict` + a lock file update to actually get this to run on our CI.

However, I think we should investigate failures early in scikit-learn because I suspect that some (most?) of them are not necessarily a problem in scikit-learn but might be bugs in `array-api-strict`'s device support itself.

/cc @betatim

ogrisel committed:
TST enable non-CPU device testing via array-api-strict

betatim commented:
I think we will need https://github.com/data-apis/array-api-strict/pull/73 and https://github.com/data-apis/array-api-strict/pull/72 for this PR to work

betatim commented:
Another issue that needs resolving https://github.com/scipy/scipy/issues/21736

betatim committed:
Use correct device when creating arrays

ogrisel committed:
Merge branch 'main' into multi-device-array-api-strict

ogrisel committed:
Make check_array_api_metric array-api-strict aware

ogrisel committed:
Similar fix in check_array_api_input

ogrisel committed:
Forward device info in temp array in LDA

ogrisel commented:
@betatim I started to update this PR: it discovered several device handling issues and maybe dtype related issues. I have not yet fixed them all, feel free to take over at any point :)

ogrisel committed:
Merge branch 'main' into multi-device-array-api-strict

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -93,6 +93,7 @@ def _assess_dimension(spectrum, rank, n_samples):
                 (spectrum[i] - spectrum[j]) * (1.0 / spectrum_[j] - 1.0 / spectrum_[i])
             ) + log(n_samples)
 
+    pu = xp.asarray(pu, device=spectrum.device)
```
`pu` is a scalar Python value (`-rank * log(2.0)` where `rank` is an `int`) so moving it to the device explicitly should not be necessary.

ogrisel replied:
If we specify the device we should probably also specify the `dtype` to avoid the automatic cast to `float64`.

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -1369,9 +1372,10 @@ def _mean_tweedie_deviance(y_true, y_pred, sample_weight, power):
+            / ((1 - p) * (2 - p))
+            - y_true * xp.pow(y_pred, xp.asarray(1 - p, device=y_true.device)) / (1 - p)
+            + xp.pow(y_pred, xp.asarray(2 - p, device=y_true.device)) / (2 - p)
         )
```
Same comments here: for scalar values we should not need to move to the device explicitly, but if we do we should also set the dtype explicitly.

OmarManzoor replied:
Is the above comment applicable i.e. do we need to set the device for scalars and if so should we consider specifying the dtype? Basically is this only a requirement for array-api-strict as I think we weren't really encountering issues with scalars for Pytorch and Cupy?

betatim replied:
I think we need to explicitly convert scalars to `xp` arrays so that we can specify the device those arrays should be placed on. This is because array-api-strict is more strict about this than other libraries.

The array API standard has been updated to allow scalars for a lot of things, so that you don't have to explicitly create arrays. But it will take some time before that change is formalised and arrives in all the libraries/we can rely on it.

With changing to use `lgamma` we have less `xp.asarray` stuff, which is nice.

OmarManzoor replied:
Yes I think that makes sense as array-api-strict has been the main limitation with these kinds of situations. In any case do you think we should specify the `dtype` as well in the above `xp.asarray` statements?

betatim replied:
I don't know :-/ Naively it seems like "choose what ever dtype you want" is fine here? In particular because at the end of the function we convert to a Python `float` anyway. But maybe it matters to numerical precision??

ogrisel replied:
Sorry, I started a new related thread by trying to suggest a change on those particular lines:

https://github.com/scikit-learn/scikit-learn/pull/30090#discussion_r1979839022

betatim committed:
Propagate device and dtype when we create new, derived arrays

betatim committed:
Switch back to using numpy/pure python for memory computation

To compute batch sizes and memory sizes we don't need to use the array
API, we can do that math with "just" Python types.

This change also fixes a slicing error that only appears with
array-api-strict. Unrelated to changing to Python types.

betatim committed:
Add custom xlogy implemntation

The scipy implementation contains a bug with respect to setting the
device of all the arrays it creates. This adds xlogy() to our group of
functions we implement ourselves.

betatim committed:
Make sure to propagate device information when creating adhoc arrays

betatim committed:
Add xp= keyword argument to get_namespace_and_device

Using this in functions that support the xp short circuiting, so I think
it makes sense to make this function look similar to get_namespace

betatim committed:
Fix

betatim committed:
Fix

betatim commented:
@ogrisel I think things are working now. Want to have a look? Most of the change swere indeed about passing `device` to array creation functions

ogrisel commented on sklearn/decomposition/_pca.py:
```diff
@@ -93,6 +93,7 @@ def _assess_dimension(spectrum, rank, n_samples):
                 (spectrum[i] - spectrum[j]) * (1.0 / spectrum_[j] - 1.0 / spectrum_[i])
             ) + log(n_samples)
 
+    pu = xp.asarray(pu, device=device(spectrum), dtype=spectrum.dtype)
```
Is this line still required? Can't we just use a regular Python scalar here? No big deal though.

betatim replied:
Yes and no. We need it, unless we change from using `scipy.special.gammaln` to `math.lgamma` on L74. With the scipy `gammaln` we get an array with a single element. And if that array is on a "GPU device" then we can't implicitly convert it to a numpy array (which is what `pu` is).

Overall I think it makes sense to just use `math.lgamma`, as we don't want to compute on arrays and even the scipy docs say that they are the same.

OmarManzoor replied:
I think either option should be fine.

ogrisel reviewed (commented):
Thanks Tim! LGTM but I cannot approve my own PR.

betatim commented:
I pinged Omar and Guillaume for reviews. You don't have to review this, but I thought it might be interesting for you two (and solve the problem that neither Oliver nor I can approve this).

OmarManzoor commented on sklearn/utils/_array_api.py:
```diff
@@ -82,6 +82,19 @@ def yield_namespace_device_dtype_combinations(include_numpy_namespaces=True):
+                yield array_namespace, array_api_strict.Device("device1"), "float32"
+            except ImportError:
+                # Those combinations will typically be skipped by pytest if
+                # array_api_strict is not installed but we still to see them in
```
```suggestion
                # array_api_strict is not installed but we still need to see them in
```

OmarManzoor reviewed (commented):
Thank you @betatim and @ogrisel. 
Generally looks good, just a few comments

betatim committed:
Switch to using math.lgamma

This reduces the amount of `xp.asarray` that we need to convert scalars
to arrays for the array API

betatim committed:
Update sklearn/utils/_array_api.py

Co-authored-by: Omar Salman <omar.salman2007@gmail.com>

OmarManzoor commented on sklearn/metrics/_regression.py:
```diff
@@ -1369,9 +1376,9 @@ def _mean_tweedie_deviance(y_true, y_pred, sample_weight, power):
-            xp.pow(y_true, xp.asarray(2 - p)) / ((1 - p) * (2 - p))
-            - y_true * xp.pow(y_pred, xp.asarray(1 - p)) / (1 - p)
-            + xp.pow(y_pred, xp.asarray(2 - p)) / (2 - p)
+            xp.pow(y_true, xp.asarray(2 - p, device=device_)) / ((1 - p) * (2 - p))
+            - y_true * xp.pow(y_pred, xp.asarray(1 - p, device=device_)) / (1 - p)
+            + xp.pow(y_pred, xp.asarray(2 - p, device=device_)) / (2 - p)
```
I think we can report but considering that using an array with the correct device works, maybe we can merge this PR so that it is not blocked while waiting for the required fix in array-api-strict. Maybe we can add a Todo to remove the additional code once this is fixed. What do you think?

ogrisel replied:
Actually, I tried again locally, and it appears that we can simplify this code as done in 920932f.

OmarManzoor replied:
Tests seem to be failing

ogrisel replied:
My bad, I switched env and used an old version of array-api-strict that did not trigger this particular problem. Will revert the last commit.

ogrisel replied:
Actually, there are two sets of changes in 920932f: some about `xp.pow` and the other about `xp.where`.

It appears that the simplification of the `xp.pow` calls is fine (I will push a new commit).

I will then check whether the scalar/device in `xp.where` is a bug or not and open an issue upstream accordingly.

ogrisel replied:
Ok I think I understand the source of the discrepancy between my local tests and the CI: by default `array_api_strict` is configured to implement the latest version of the spec ("2023.12") but our CI infrastructure is using "2023.12".

We should probably bump this and simplify the code accordingly.

ogrisel committed:
Remove xp.asarray(..., device=device_) idioms in _mean_tweedie_deviance

ogrisel committed:
Revert "Remove xp.asarray(..., device=device_) idioms in _mean_tweedie_deviance"

This reverts commit 920932faacae6a96789a73c6b824d15d31969c07.

ogrisel committed:
set_array_api_strict_flags(api_version="2024.12")

ogrisel committed:
Directly pass Python scalars to _mean_tweedie_deviance

ogrisel committed:
Merge branch 'main' into multi-device-array-api-strict

ogrisel committed:
Missing dtype in _nanmean & consistent xp.asarray arg order

OmarManzoor commented:
> @OmarManzoor @betatim after [#30090 (comment)](https://github.com/scikit-learn/scikit-learn/pull/30090#discussion_r1983414399), the code is simpler, and all tests pass everywhere.
> 
> +1 for merge on my side.

👍 Let's wait for the CI to complete and I'll review and merge

OmarManzoor commented on sklearn/metrics/pairwise.py:
```diff
@@ -621,13 +618,13 @@ def _euclidean_distances_upcast(X, XX=None, Y=None, YY=None, batch_size=None):
     xp_max_float = _max_precision_float_dtype(xp=xp, device=device_)
     for i, x_slice in enumerate(x_batches):
-        X_chunk = xp.astype(X[x_slice], xp_max_float)
+        X_chunk = xp.astype(X[x_slice, :], xp_max_float)
```
Why do we need this exactly?

ogrisel replied:
Otherwise we get:

```
IndexError: self.ndim=2, but the multi-axes index only specifies 1 dimensions. If this was intentional, add a trailing ellipsis (...) which expands into as many slices (:) as necessary - this is what np.ndarray arrays implicitly do, but such flat indexing behaviour is not specified in the Array API.
```

OmarManzoor replied:
I see. Thanks for clarifying.

OmarManzoor commented on sklearn/utils/_array_api.py:
```diff
@@ -769,49 +792,66 @@ def _average(a, axis=None, weights=None, normalize=True, xp=None):
+
+    with numpy.errstate(divide="ignore", invalid="ignore"):
+        temp = x * xp.log(y)
+    return xp.where(x == 0.0, xp.asarray(0.0, dtype=temp.dtype, device=device_), temp)
```
Can't we simplify the arrays that we are creating in this function as well as in the `_nanmin, _nanmax, _nanmean`

ogrisel replied:
I tried, but passing an explicit `dtype` is mandatory to avoid upcasting to `float64` instead of keeping computation happening in `float32`. For instance, it would break `MaxAbsScaler` because the `.scale_` parameter would be `float64` when fit on `float32` data.

OmarManzoor replied:
I see. Yes that makes sense.

OmarManzoor reviewed (approved):
LGTM. Thank you @ogrisel and @betatim
